From 828df4d292f2ad0eeafad7d8e010c497c9a4ee71 Mon Sep 17 00:00:00 2001
From: Val Packett <val@invisiblethingslab.com>
Date: Thu, 27 Aug 2026 06:38:42 -0300
Subject: [PATCH] drm/virtio: use CREATE_GUEST_HANDLE for imported DMA-BUFs

When the CREATE_GUEST_HANDLE feature has been negotiated, automatically
create host-side handles for resources imported from external devices.

There is no way for userspace to explicitly control this flag during the
PRIME import operation, but the assumption is reasonable because there
is unlikely to be a use case for imported buffers that would not benefit
from host-side handles. So far, imported buffers have only been used
with KMS/"2D"-only virtio-gpu setups, for passing buffers rendered on
a passed-through physical GPU to a host VMM's display for scanout.
Now it is also possible to use such buffers, or ordinary guest memfds
converted to udmabufs, in a cross-domain context to send to the host
Wayland compositor, which requires a host-side handle.

Signed-off-by: Val Packett <val@invisiblethingslab.com>
---
 drivers/gpu/drm/virtio/virtgpu_prime.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/drivers/gpu/drm/virtio/virtgpu_prime.c b/drivers/gpu/drm/virtio/virtgpu_prime.c
index 216c77cd0d21..cf8454a47353 100644
--- a/drivers/gpu/drm/virtio/virtgpu_prime.c
+++ b/drivers/gpu/drm/virtio/virtgpu_prime.c
@@ -255,6 +255,10 @@ static int virtgpu_dma_buf_init_obj(struct drm_device *dev,
 	params.blob = true;
 	params.blob_mem = VIRTGPU_BLOB_MEM_GUEST;
 	params.blob_flags = VIRTGPU_BLOB_FLAG_USE_SHAREABLE;
+	if (vgdev->has_create_guest_handle)
+		/* No way for userspace to decide with this API, but most existing
+		 * use cases do require shareable handles on the host side */
+		params.blob_flags |= VIRTGPU_BLOB_FLAG_CREATE_GUEST_HANDLE;
 	params.size = attach->dmabuf->size;
 
 	virtio_gpu_cmd_resource_create_blob(vgdev, bo, &params,
